cephfs/path_test.go

```diff
@@ -60,3 +60,300 @@ func TestRemoveDir(t *testing.T) {
 	assert.EqualError(t, err,
 		fmt.Sprintf("stat %s: no such file or directory", localPath))
 }
+
+func TestLink(t *testing.T) {
+	mount := fsConnect(t)
+	defer fsDisconnect(t, mount)
+
+	t.Run("rootDirOperations", func(t *testing.T) {
+		// Root dir, both as source and destination.
+		err := mount.Link("/", "/")
+		// Error directory operations are not allowed.
+		assert.Error(t, err)
+
+		dir1 := "myDir1"
+		assert.NoError(t, mount.MakeDir(dir1, 0755))
+		defer func() {
+			assert.NoError(t, mount.RemoveDir(dir1))
+		}()
+
+		// Creating link for a directory.
+		err = mount.Link(dir1, "/")
+		// Error, directory operations not allowed.
+		assert.Error(t, err)
+	})
+
+	// Non-root directory operations.
+	fname := "testFile.txt"
+	dir2 := "myDir2"
+	assert.NoError(t, mount.MakeDir(dir2, 0755))
+	defer func() {
+		assert.NoError(t, mount.RemoveDir(dir2))
+	}()
+
+	t.Run("dirAsSource", func(t *testing.T) {
+		err := mount.Link(dir2, fname)
+		// Error, directory operations not allowed.
+		assert.Error(t, err)
+	})
+
+	t.Run("dirAsDestination", func(t *testing.T) {
+		f1, err := mount.Open(fname, os.O_WRONLY|os.O_CREATE, 0666)
+		assert.NotNil(t, f1)
+		assert.NoError(t, err)
+		defer func() {
+			assert.NoError(t, f1.Close())
+			assert.NoError(t, mount.Unlink(fname))
+		}()
+		err = mount.Link(fname, dir2)
+		// Error, destination exists.
+		assert.Error(t, err)
+	})
+
+	// File operations.
+	t.Run("sourceDoesNotExist", func(t *testing.T) {
+		fname := "notExist.txt"
+		err := mount.Link(fname, "hardlnk")
+		// Error, file does not exist.
+		assert.Error(t, err)
+	})
+
+	t.Run("sourceExistsSuccess", func(t *testing.T) {
+		useMount(t)
```

Comment:
inconsistent use of `useMount()`, should that not be called only once at the beginning of each `func Test*()`?

Reply:
No blocking issues from my point of view, I am happy the way this currently is. I am fine with changes done in follow-up PRs.

Reply:
The `useMount` functions should be called if-and-only-if the test (or subtest) makes use of the local (fuse based) cephfs mount, instead of doing everything through the API. Depending on env vars it will either skip the test (or subtest) or fail the test (or subtest). The former is the default and meant for the developer's desktop type use case and the latter is meant for the CI.
When I do a 2nd pass of the review I'll try to look for cases where this is either missing or not needed, but I am only human. :-)

Reply:
Yeah, I have also used it where I am calling Stat() which needs CephMountDir. But instead of calling it once inside the function I have called it per sub-test if that test case is calling Stat()

IMHO it shouldn't be a problem and can be fixed in separate PR, let me know if you think otherwise else we can merge this PR.